packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -336,10 +336,13 @@ export function esbuildCjsExternalPlugin(
       build.onLoad(
         { filter: /.*/, namespace: cjsExternalFacadeNamespace },
         (args) => ({
-          contents:
-            `import * as m from ${JSON.stringify(
-              nonFacadePrefix + args.path,
-            )};` + `module.exports = { ...m };`,
+          contents: `\
+import * as m from ${JSON.stringify(nonFacadePrefix + args.path)};
+if (typeof m.default === 'function') {
+  module.exports = m.default;
+} else {
+  module.exports = { ...m };
+}`,
```

Comment:
Thanks for the change! I realize my earlier comment might have been misleading, sorry about that. What I had in mind was to apply the change _only_ for Node builtins. So the change I was thinking of looks like this:
```suggestion
          contents:  isNodeBuiltin(args.path) ? `\
import * as m from ${JSON.stringify(nonFacadePrefix + args.path)};
if (typeof m.default === 'function') {
  module.exports = m.default;
} else {
  module.exports = { ...m };
}` : `\
import * as m from ${JSON.stringify(
  nonFacadePrefix + args.path,
)};` + `module.exports = { ...m };`,
```

Reply:
Ah, OK, thanks. I was curious if we could always just use the default export for Node builtins. This modified version of your script confirms this would work.

```js
import module from "node:module";

const require = module.createRequire(import.meta.url);

for (const mod of module.builtinModules) {
  const esmExport = await import(mod);
  const moduleExports = require(mod);

  const esmList = Object.keys(esmExport.default);
  const cjsList = Object.keys(moduleExports);
  const diffOnlyEsm = esmList.filter((key) => !cjsList.includes(key));
  const diffOnlyCjs = cjsList.filter((key) => !esmList.includes(key));

  console.assert(
    diffOnlyEsm.length === 0 && diffOnlyCjs.length === 0,
    mod,
    diffOnlyEsm
      .map((key) => `+ ${key}`)
      .concat(diffOnlyCjs.map((key) => `- ${key}`))
  );
}
```

Shall we go with that as it's simpler?

Reply:
Ah, that's true. I think we can go with your one.

Reply:
I've made the changes - [0c0e387](https://github.com/vitejs/vite/pull/20048/commits/0c0e387cbf750a6fe3d18cf30e7fd6d74011ffd3). Hopefully that's enough to solve the issues here.

It looks like the `getAugmentedNamespace` function used in `@rollup/plugin-commonjs` is quite a bit cleverer in how it handles these things. Out of interest, will Rolldown in Vite means that all this behaviour becomes more aligned in dev and build?